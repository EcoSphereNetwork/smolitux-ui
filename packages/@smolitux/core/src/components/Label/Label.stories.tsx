import type { Meta, StoryObj } from '@storybook/react';
import { Label } from './Label';

const meta: Meta<typeof Label> = {
  title: 'Components/Core/Label',
  component: Label,
  tags: ['autodocs'],
  argTypes: {
    children: { control: 'text' },
    htmlFor: { control: 'text' },
    required: { control: 'boolean' },
    disabled: { control: 'boolean' },
    size: { control: 'inline-radio', options: ['sm', 'md', 'lg'] },
    variant: { control: 'inline-radio', options: ['default', 'subtle', 'strong'] },
  },
  args: {
    children: 'Label text',
  },
};
export default meta;

type Story = StoryObj<typeof Label>;

export const Default: Story = {};

export const Required: Story = {
  args: { required: true },
};

export const Disabled: Story = {
  args: { disabled: true },
};

export const Sizes: Story = {
  render: (args) => (
    <div style={{ display: 'flex', gap: '1rem' }}>
      <Label {...args} size="sm" />
      <Label {...args} size="md" />
      <Label {...args} size="lg" />
    </div>
  ),
  args: { children: 'Size demo' },
};
