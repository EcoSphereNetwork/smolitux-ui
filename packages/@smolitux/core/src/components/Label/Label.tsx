import React, { forwardRef, isValidElement, cloneElement } from 'react';
import { cn } from '@smolitux/utils';

export interface LabelProps extends React.LabelHTMLAttributes<HTMLLabelElement> {
  /** Label content */
  children: React.ReactNode;
  /** htmlFor attribute */
  htmlFor?: string;
  /** Mark field as required */
  required?: boolean;
  /** Mark as disabled */
  disabled?: boolean;
  /** Font size */
  size?: 'sm' | 'md' | 'lg';
  /** Visual variant */
  variant?: 'default' | 'subtle' | 'strong';
  /** Render child element instead of label */
  asChild?: boolean;
}

export const Label = forwardRef<HTMLLabelElement, LabelProps>(
  (
    {
      children,
      htmlFor,
      required = false,
      disabled = false,
      size = 'md',
      variant = 'default',
      asChild = false,
      className,
      ...rest
    },
    ref,
  ) => {
    const classes = cn(
      'smx-label',
      {
        'text-sm': size === 'sm',
        'text-base': size === 'md',
        'text-lg': size === 'lg',
        'font-semibold': variant === 'strong',
        'text-gray-700': variant === 'subtle',
      },
      disabled && 'opacity-50 cursor-not-allowed',
      className,
    );

    if (asChild && isValidElement(children)) {
      return cloneElement(children as React.ReactElement, {
        ref,
        htmlFor,
        'aria-disabled': disabled || undefined,
        'aria-required': required || undefined,
        className: cn(classes, (children as any).props.className),
        ...rest,
      });
    }

    return (
      <label
        ref={ref}
        htmlFor={htmlFor}
        className={classes}
        aria-disabled={disabled || undefined}
        aria-required={required || undefined}
        {...rest}
      >
        {children}
        {required && (
          <span className="text-red-500 ml-1" aria-hidden="true">
            *
          </span>
        )}
      </label>
    );
  },
);

Label.displayName = 'Label';
export default Label;
