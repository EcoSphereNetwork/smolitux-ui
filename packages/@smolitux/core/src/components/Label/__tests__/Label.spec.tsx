import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Label } from '../Label';

describe('Label', () => {
  test('snapshot', () => {
    const tree = renderer
      .create(<Label htmlFor="x">Name</Label>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders htmlFor', () => {
    render(<Label htmlFor="field">Label</Label>);
    const label = screen.getByText('Label');
    expect(label).toHaveAttribute('for', 'field');
  });

  test('applies aria attributes', () => {
    render(
      <Label htmlFor="f" required disabled>
        L
      </Label>,
    );
    const label = screen.getByText('L');
    expect(label).toHaveAttribute('aria-required', 'true');
    expect(label).toHaveAttribute('aria-disabled', 'true');
  });
});
