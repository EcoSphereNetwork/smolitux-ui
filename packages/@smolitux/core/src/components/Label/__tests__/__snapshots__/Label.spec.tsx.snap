// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label snapshot 1`] = `
<label
  className="smx-label text-base"
  htmlFor="x"
>
  Name
</label>
`;
